ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+    "distributed.scheduler.allowed-failures": 0,
+    # FSDP forward/backward passes can take way longer than the default warning at 3s
+    "distributed.admin.tick.limit": "30s",
+    # Gives us more time to debug a downed worker. TODO: see if there are negative consequences
+    # of having this always enabled
+    "distributed.comm.timeouts.connect": "300s",
+    "distributed.comm.timeouts.tcp": "300s",
+})
+
+logger = logging.getLogger(__name__)
+
+TReturn = TypeVar("TReturn")
+TParams = ParamSpec("TParams")
+
+
+def is_conversation(messages) -> bool:
+    """Check if messages is an instance of Conversation."""
+    return isinstance(messages, list) and all(
+        isinstance(msg, dict)
+        and all(
+            isinstance(key, str) and isinstance(value, str)
+            for key, value in msg.items()
+        )
+        for msg in messages
+    )
+
+
+def get_unused_port() -> int:
+    """Find an unused port by creating a temporary socket."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
+        return s.getsockname()[1]
+
+
+class ExecutionMode(StrEnum):
+    LOCAL_MACHINE = auto()
+    SLURM_CLUSTER = auto()
+
+
+class FSDPConfig(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    offload_cpu: bool = Field(
+        default=False,
+        description="Whether to offload model parameters and gradients to CPU.",
+    )
+    activation_checkpointing: bool = False
+    cpu_ram_efficient_loading: bool = False
+    state_dict_type: StateDictType = StateDictType.FULL_STATE_DICT
+    backward_prefetch: BackwardPrefetch | None = Field(
+        BackwardPrefetch.BACKWARD_PRE,
+        description=(
+            "See https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.BackwardPrefetchDefault"
+            " is PRE to be consistent with FSDP's default."
+        ),
+    )
+
+    @field_validator("backward_prefetch", mode="before")
+    @classmethod
+    def validate_backward_prefetch(cls, v) -> BackwardPrefetch | None:
+        if v is None:
+            return None
+        if isinstance(v, str):
+            # BackwardPrefetch isn't a StrEnum, so Pydantic won't cast a string
+            return BackwardPrefetch[v]
+        return BackwardPrefetch(v)
+
+
+class ParallelModeConfig(FSDPConfig):
+    num_workers: int = Field(description="Number of workers to be used in the cluster.")
+    num_cpus_per_worker: int = Field(
+        default=1, description="Number of CPUs to be allocated per worker."
+    )
+    execution_mode: ExecutionMode = Field(
+        description=(
+            "Execution mode of the current setup, defines how to allocate resources"
+            " (cpus/gpus) for the model to run on."
+        ),
+        default=ExecutionMode.LOCAL_MACHINE,
+    )
+
+    scheduler_addr: str = "localhost"
+    scheduler_port: int = Field(default=0, description="0 means Dask picks randomly.")
+    torch_port: int = Field(default_factory=get_unused_port)
```

Comment:
Can you add a `description` to what `torch_port` means? To me, `torch` is a framework, not an OS.

Maybe just do `ParallelWorkerConfig.model_fields["xyz"].description` to make the linkage clear

So maybe we can make the name more verbose, or add a description stating the implications

Reply:
I don't understand the framework/OS comment, nor the `model_fields` suggestion.

In any case, I think `torch_port` is clear - it's the port that torch uses to communicate between processes. Fairly standard configuration that should be familiar to users who understand torch distributed.